### ISSUE TITLE
Idea: doFetch returns a promise

### ISSUE THREAD

**Issue by shanecav** [jamesplease/react-request#146]
What do you think about having `doFetch` return a promise that resolves with the same object that the `afterFetch` function receives?

My use case is that I have a form that's a child of two composed `react-request` `Fetch` components (one to GET the initial form values, and one to PUT the updated values). I need the form's submission handler to call `myPutFetch.doFetch()` and then call a `setSubmitting(false)` function (which is provided as an arg in the submission handler) when the fetch is complete. I also would like to be able to have that submission handler have access to the PUT fetch's response when it's complete, so I can show errors from the server, etc.

So I'd like to be able to do something like this:

```js
<MyComposedFetches>
  {([ myGet, myPut ]) => (
    <MyForm
      initialValues={myGet.data}
      onSubmit={(values, { setSubmitting }) => {
        myPut.doFetch({ body: JSON.stringify(values) })
          .then(({ failed, error, didUnmount }) => {
            setSubmitting(false);
            if (!didUnmount && !failed && !error) {
              alert('success');
              myGet.doFetch();
            }
          })
      }}
    />
  )}
</MyComposedFetches>
```

Hopefully that makes sense. Let me know what you think - thanks!

**Comment by jamesplease** [comment]
I see what you’re trying to do, @shanecav . I think a change is needed somewhere, possibly the one you proposed. I’m going to need to put more thought into this one, though 🤔

**Comment by shanecav** [comment]
Hey @jamesplease, any new thoughts on this? I'm happy to help out however I can, just let me know - thanks!

**Comment by jamesplease** [comment]
Haven’t had time to think about it, sorry! I’ll post an update here once I have one

**Comment by jamesplease** [comment]
As a workaround @shanecav , would something like this work for you?

1. Store the `failed`, `error`, and `didUnmount` state in the component's state.
2. Access that state in the callback

In this code, I didn't add in the `afterFetch` call that would set the state, but it could be consumed like the following:

```jsx
<MyComposedFetches>
  {([ myGet, myPut ]) => (
    <MyForm
      initialValues={myGet.data}
      onSubmit={(values, { setSubmitting }) => {
        myPut.doFetch({ body: JSON.stringify(values) })
          .then(() => {
            // pull this off of state
            const { failed, error, didUnmount } = this.state;
            setSubmitting(false);
            if (!didUnmount && !failed && !error) {
              alert('success');
              myGet.doFetch();
            }
          })
      }}
    />
  )}
</MyComposedFetches>
```

As an aside, I admit I am a little hesitant to make big changes to how the requests work here, mostly because of React Suspense. I think of changing the signature of `doFetch` to be a pretty big change, so I would love if we could find an alternative way for you to do what you are trying to do.

If the above system doesn't work for you, maybe we could set up time to do a GHangout to go over what you are trying to do, and what the different options are?

Thanks!

**Comment by shanecav** [comment]
Unless I'm mistaken, I can't do `myPut.doFetch(...).then(...)`, since `doFetch` doesn't return a promise (or anything else) currently - correct?

I'm currently doing a similar work-around though: in my form's `onSubmit` handler, I set some state in an ancestor component, which I can then access from the Fetch component's `afterFetch` prop callback. It's working fine for now, just seems a little dirty 😃.

Totally understand about React Suspense anticipation though, and I definitely appreciate the care around making substantial API changes. My general goal is to allow the caller of `doFetch` to be able to react to the fetch's completion. If there's a better way to accomplish that, I'm all for it!

Thanks @jamesplease!

**Comment by jamesplease** [comment]
> Unless I'm mistaken, I can't do myPut.doFetch(...).then(...), since doFetch doesn't return a promise (or anything else) currently - correct?

Ah, you are right.

I think that should at least be updated to return the native promise from `fetch`. What do you think?

> I'm currently doing a similar work-around though: in my form's onSubmit handler, I set some state in an ancestor component, which I can then access from the Fetch component's afterFetch prop callback. It's working fine for now, just seems a little dirty 😃.

Makes sense. I see why that would feel dirty. If `doFetch` returned a Promise – even if it doesn't have the same data that you have in `afterFetch`, would that clean up your code at all, or no?

**Comment by shanecav** [comment]
> I think that should at least be updated to return the promise from fetch. What do you think?

That would work, too. I initially suggested the same signature as `afterFetch` since it seems like `doFetch`'s "completion" serves the same purpose, and that extra data is nice to have. Are you trying to keep `doFetch` more inline with the `fetch` API?

> If doFetch returned a Promise – even if it doesn't have the same data that you have in afterFetch, would that clean up your code at all, or no?

Yes, definitely. I'd have to use both `afterFetch` and `doFetch.then` simultaneously in order to cover everything I need to do, but at least I'd be able to stop manually managing that data through an ancestor component's state. That alone would be a big improvement.

**Comment by jamesplease** [comment]
> Are you trying to keep doFetch more inline with the fetch API?

Yeah, that's what I was thinking, although it may not be a good idea.

I have to apologize, because I was mistaken going into this issue. I was under the impression that `doFetch` already returned a Promise that matched the `fetch` API. A big change, in my opinion, would have been going from _that_ to the `afterFetch` signature. But since it returns nothing right now, that opens the door for us to pick whichever one that we think is best.

> That alone would be a big improvement.

Okay, awesome.

At this point, I am 100% certain that `doFetch` should return a Promise. I am not entirely certain if it should act like a `fetch()` wrapper, or use the same `afterFetch` signature. I can see it going both ways.

Do you have any more arguments to make in favor it following the `afterFetch` signature? If not, no big deal. I'll spend a few more days thinking about this, and by this weekend I think we could get something released. Does that timeline work for you?

Thanks for your patience and help with this, @shanecav ! I appreciate it! And sorry again for the delayed response.

**Comment by shanecav** [comment]
The arguments I see in favor of the `afterFetch` signature are:
1. There wouldn't be a need to use both `doFetch().then(...)` and `afterFetch` simultaneously (in order to access the extra data `afterFetch` provides that `fetch`'s API doesn't). Having to use both in parallel would be a bit complex, and it wouldn't be obvious to someone reading the `doFetch().then(...)` code that the request is also being responded to elsewhere (in some parent Fetch's `afterFetch` prop).
2. IMO it would be easier to understand/remember that `doFetch`'s promise resolution and `afterFetch` use the same signature. I'd have an easier time thinking "this is how `react-request` always handles completed requests" as opposed to having to remember which one uses which signature.
3. The render prop and `afterFetch` already set a precedent for expanding on the standard `fetch` API for the sake of added convenience. I personally see `doFetch` as falling in the same category.

And no apology necessary, I'm just grateful you published and are supporting this awesome library! I'm happy with any timeline, and am more than happy to help out with PR/docs/whatever 😃 Thanks for your time on this!

**Comment by jamesplease** [comment]
@shanecav , I am working hard on fixing #151 before returning back here, but I haven't forgotten about ya! This bug fix has taken a lot of time to do, which is why I haven't gotten around to this yet.

Sorry about the delay!!

**Comment by shanecav** [comment]
No worries, and thanks for the update!

**Comment by jamesplease** [comment]
@shanecav thanks for sharing your arguments for `doFetch` following `afterFetch`’s signature.

1. Agreed. Using both sounds really bad.
2. Agreed, I can see how this would be easier for someone to remember.
3. Interestingly, I don’t think of `afterFetch` nor the render prop as expanding on the standard fetch API. I have just thought of it as being _the_ API that is a necessity for the React lifecycle (`fetch-dedupe`, on the other hand, is what I think of as expanding on the standard fetch API).

I asked some folks at work what they think. One good argument I heard is that if `doFetch` returns the `afterFetch` arguments, then folks could be confused about which to use.

I’m still not decided either way. I can see it going either way. I think I need to think more about, and optimize, for the most common use case for this, which I think is something like your use case. At the moment that I am typing this, I am leaning toward following the `afterFetch` API, but a few days ago I was thinking of making it return the regular Promise from `fetch-dedupe` so...we’ll see.

Sorry for the slowness here, btw. Still focused on that bug (which I believe is fixed. We’re going to be testing it at work today before I cut a release). This, and updating to `fetch-dedupe` for your awesome `responseType` fixed are my main priorities after the bug fix lands!

**Comment by shanecav** [comment]
> Interestingly, I don’t think of afterFetch nor the render prop as expanding on the standard fetch API. I have just thought of it as being the API that is required to map a request to the React lifecycle (fetch-dedupe, on the other hand, is what I think of as expanding on the standard fetch API).

That's a much better way of putting it! 😄 

> One good argument I heard is that if doFetch returns the afterFetch arguments, then folks could be confused about which to use. This could be resolved with good documentation stating the right and wrong way to use both, though.

I see the difference in use cases between `doFetch`'s returned Promise vs. `afterFetch` as similar to the difference between `doFetch`'s options argument vs. the `Fetch` props that they can override: It comes down to whether you need access to data in the child scope where you call `doFetch`.

A form is a great example of where I need access to data in a child scope when making & responding to a request, and where the `afterFetch` API in `doFetch`'s returned Promise would be extra helpful:

1. When I call `doFetch`, I include the form values in the `body` of the request via `doFetch`'s options argument.
2. I then need to respond to that request by setting the form's `submitting` property to false, and potentially set some form field errors if the server returned any. This is where the `afterFetch` API would be really helpful - it makes it easy to do things like checking if it `failed` and then setting form field errors based on `data`.

Thank you for the work and thought you're putting into this, by the way - it's very much appreciated! And again, happy to help out however I can with implementation or docs - just let me know what I can do 😄.

**Comment by jamesplease** [comment]
@shanecav , are you thinking that the `doFetch` promise will always resolve with the after fetch props, and never reject?

i.e.;

```js
doFetch({ ...stuff })
  .then(({ error }) => {
    if (failed) {
      console.log('ya it failed', failed);
    }
  })
```

**Comment by shanecav** [comment]
Yeah, that's what I was thinking. My understanding is that's how `afterFetch` currently works: if the underlying `fetch` promise rejects, `afterFetch` still gets called, and the `error` property of its argument comes from that `fetch` rejection - is that correct?

I'd personally expect the `doFetch` promise resolution to behave the same way if it's receiving the same argument as `afterFetch` - i.e. it's basically the same as `afterFetch`, it just lives in a different scope.

But maybe there's some reason for wanting a promise rejection that I haven't thought of. Can you think of any?

**Comment by jamesplease** [comment]
@shanecav , I am going to go ahead and release 3.0.0 with the `responseType` change, but without this. I agree with this issue for two separate reasons:

1. `doFetch` _absolutely_ needs to return a Promise
2. the serial fetch story is not clearly outlined ( #59 )

Serial fetches is one thing I have been meaning to sit down and spend more time thinking about, but I just haven't had the time. Some folks at work and I are going to spend time in the next week deciding what we think the best approach is.

There will likely be two stories: one for the `doFetch()` use case, and the other for the props use case (as these are the two ways you can make a request using this lib).

Sorry for the extra delay...it's just that I want to make sure this is implemented right 🙈 and, as silly as it sounds, it wasn't until a few days ago that I made the connection between this issue and #59 😓

**Comment by ianstormtaylor** [comment]
Just wanted to +1 this, it would be great for handling success/failure function props that the parent has passed down.

**Comment by jamesplease** [comment]
Let’s go with the `afterFetch` signature. Is anyone up for making a PR?

**Comment by shanecav** [comment]
Happy to give it a go, but probably won’t have time for a week or two.

**Comment by jamesplease** [comment]
Resolved in #183 ; released as 3.1.0

**Comment by shanecav** [comment]
Awesome! Happy to help, and thank you for the time and thought you put into this issue!
